File: twopt.jl
using QuadGK
using SpecialFunctions
using DelimitedFiles
using Trapz
using Plots
using Interpolations

# Define cosmological parameters for DUSTGRAIN-pathfinder simulations
# as they appear in https://doi:10.1093/mnras/sty2465 for LCDM background
Ω_b = 0.0481
h = 0.6731
H0 = h*100 # Hubble constant in km/s/Mpc

# For fr4
Ω_m = 0.31345
Ω_cdm = Ω_m - Ω_b
Ω_de = 1 - Ω_m

plot_label = "fR4"

# Load matter power spectrum from file (assuming it contains only P_kappa(ell) values)
Pnl = readdlm("fr4.txt")
k = readdlm("fr4_k.txt")
z = readdlm("fr4_z.txt")

zs = 0.5

# Function to calculate E(z)
function E(z)
    return sqrt(Ω_m * (1 + z)^3 + Ω_de)
end

# Function to calculate comoving distance r(z)
function r(z)
    c = 299792.458 # Speed of light in km/s
    integrand(z_prime) = 1 / E(z_prime)
    result, _ = quadgk(integrand, 0.0, z)
    return (c / H0) * result
end

zk_values = (z,k)

Pzk = CubicSplineInterpolation(zk_values, Pnl)

# Angular power spectrum without tomography
function C_ℓ(ℓ)
    c = 299792.458 # Speed of light in km/s

    # Function to calculate W(z)
    function W(z)
        return 1.5 * Ω_m * (H0 / c)^2 * (1 + z) * r(z) * (1 - (r(z) / r(zs)))
    end

    # Integrating function
    integrand(z) = W(z)^2 * P_zk[z, (ℓ + 0.5) / r(z)]/ r(z)^2 / E(z)
    
    # Perform the integral using quadgk
    result, _ = quadgk(integrand, 0.0, zs)
    
    # Return the result
    return (c / H0) * result
end

# Define the range of ell values
ℓ = range(10, stop=10000, length=100)

# Define the range of theta values
θ = exp10.(range(start=0.1, stop=100, length=100))

# Define the integrand function for xi_plus(theta)
function integrand(θ, ℓ)
    return (1 / (2π)) * C_ℓ(ℓ)* besselj0(θ * ℓ) * ℓ
end


# Compute xi_plus(theta) for each theta value
ξ = zeros(length(θ))
for (i, val_θ) in enumerate(θ)
    integrand_ℓ(ℓ) = integrand(val_θ, ℓ)
    ξ[i], _ = quadgk(integrand_ℓ, ℓ)
end

# Plot xi_plus(theta)
plot(θ, ξ, xaxis=:log, yaxis=:log, xlabel="θ", ylabel="ξ_+(θ)", title="Two-point Correlation Function")

File: twopt.f90
program TwoPointCorrelation
    implicit none
    
    ! Libraries
    external besselj0
    external quadgk
    external readdlm

    ! Constants
    real, parameter :: pi = 3.14159265358979323846
    real, parameter :: theta_min = 0.01
    real, parameter :: theta_max = 100.0
    integer, parameter :: num_theta = 100
    integer, parameter :: num_ell = 100
    
    ! Arrays
    real :: theta(num_theta), xi_plus(num_theta)
    real :: ell(num_ell), P_kappa(num_ell)
    
    ! Load matter power spectrum from file
    call readdlm("\home\alessandro\phd\scripts\Dustgrain_outs\Cls\fr4_0.5")
    
    ! Define the range of ell values
    ell = [(10.0 + (100000.0 - 10.0) * real(i-1) / real(num_ell-1), i = 1, num_ell)]
    
    ! Define the range of theta values
    theta = 10.0 ** [(log10(theta_min) + (log10(theta_max) - log10(theta_min)) * real(i-1) / real(num_theta-1), i = 1, num_theta)]
    
    ! Compute xi_plus(theta) for each theta value
    do i = 1, num_theta
        xi_plus(i) = 0.0
        do j = 1, num_ell
            xi_plus(i) = xi_plus(i) + quadgk(integrand(theta(i), ell(j)), 10.0, 100000.0)
        end do
        xi_plus(i) = xi_plus(i) / (2 * pi)
    end do
    
    ! Output
    open(unit=10, file="xi_plus.txt", status="replace")
    do i = 1, num_theta
        write(10, '(2E20.10)') theta(i), xi_plus(i)
    end do
    close(10)
    
contains

    real function integrand(theta, ell)
        real, intent(in) :: theta, ell
        integrand = P_kappa(j) * besselj0(theta * ell) * ell
    end function integrand
    
end program TwoPointCorrelation
